provisioning/WFLY-18671_Galleon_tooling_upward_compat.adoc

```diff
@@ -0,0 +1,219 @@
+
+=== Relevant Installation Types
+// Remove the x next to the relevant field if the feature in question is not relevant
+// to that kind of WildFly installation
+* [x] Traditional standalone server (unzipped or provisioned by Galleon)
+
+* [ ] Managed domain
+
+* [x] OpenShift s2i
+
+* [x] Bootable jar
+
+== Requirements
+
+=== Implementation notes
+
+==== Tooling API
+
+The API is composed of 3 parts. Being 3 parts is an implementation detail, tooling depend directly on the Galleon API only.
+
+* Galleon maven universe: The resolution of Galleon feature-packs using Maven.
+
+* Galleon common API: this API is shared between Galleon core and the Galleon API.
+
+* Galleon API: Provisioning + feature-pack resolutions. It depends on the Galleon common API and Galleon maven universe.
+
+The maven coordinates of the Galleon API: `org.jboss.galleon:galleon-api", coordinates of the common api: `org.jboss.galleon:galleon-common-api`, 
+coordinates of the maven universe api: `org.jboss.galleon:galleon-maven-universe`.
+
+==== No Galleon core classes in the tooling classpath
+
+The Galleon core classes must not be present in the tooling classpath. Dynamic classloader is to be used to resolve Galleon core classes.
+
+==== Resolution of the Galleon core library
+
+At provisioning time, the new API resolves the feature-pack(s) to use and retrieve the Galleon core version advertised by the feature-pack(s).
+When multiple galleon feature-packs are used, the greater version is used. If a discovered version is smaller than the API version, the API version is used.
+If no version is discovered (feature-pack built with an older version of Galleon), the version of the API in use is used. 
+Galleon being backward compatible, older feature-packs can be provisioned with this new API.
+
+When WildFly Channels are configured in the provisioning tooling, an attempt is first made to identify if a Galleon core version is configured in the channel. 
+If a version is found, this version is used and no discovery of the Galleon core version is attempted. If no version is found, the version discovery is operated.
+
+This discovery is transparent when using the Galleon tooling API.
+
+==== Packaged Galleon core artifact
+ 
+The API jar artifact includes the Galleon core library jar artifact of the same version. N.B.: The API includes the Galleon core jar, 
+the Galleon core classes are hidden and not loadable from the classpath.
+
+Packaging the core jar allows to avoid to resolve the core library when the version is equal or smaller to the current API version.
+For other cases, the jar artifact is resolved using the configured maven repositories (according to the tooling in use).
+
+==== Impact on WildFly Galleon plugins
+
+This plugin is WildFly specific. It is called during provisioning in order to assemble a WildFly installation. This plugin is bound 
+to a given Galleon core version. The new API enforces that the expected version is used during provisioning.
```

Comment:
What is the expected version of the plugin? Version from the feature pack's descriptor?